### PR TITLE
New Article Cards

### PR DESCRIPTION
**Pull request by UladBohdan**
Article Cards redesigned and refactored.

New features:
* Toggle on Admin Preview page: switch between cards and article.
* Border for "light" cards
* Turquoise hover for all cards (Articles & Tags)
* Multiple brands on cards
* Responsive images on cards
* Images on cards do not get shown if absent (background color is used instead)
* Themes and Colors for Tag Cards

Fixes suggested by Vitalik:
* Footer copyright paddings
* Layout for very wide screens
* Diary line-spacing removed
* Article page tags (brands and authors)
* Sidebar section paddings

### CODE REVIEW

**Comment by UladBohdan** [comment]
[redesign for Article/Tag cards as described in figma](https://trello.com/c/QYIhFK6p/406-redesign-for-article-tag-cards-as-described-in-figma)

File: components/articles/Article.js [hunk 1]
```diff
@@ -60,7 +60,10 @@ const Article = ({
               </Link>
             ))}
             <div className="article-page__tag-titles">
-              <span>{renderNodeList(brands.map(tag => getTagLink({ tag })))}</span>
+              <span>
+                {renderNodeList(brands.map(tag => getTagLink({ tag })))}
+                {!!brands.length && ','}
```
Comment (Drapegnik) [inline components/articles/Article.js]: what if there are no authors?

File: components/articles/cards/ArticleCard.js [hunk 2]
```diff
@@ -42,45 +42,80 @@ const ArticleCard = props => {
...
-    );
-  }
-
   return (
-    <CardWrapper {...wrapperProps} className="article">
-      <div className="article__cover-wrapper">
-        <img className="article__cover article__cover--vertical" src={vertical} alt={title} />
-        <img className="article__cover article__cover--horizontal" src={horizontal} alt={title} />
+    <CardWrapper
+      {...wrapperProps}
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: now it's redundant, just pass props explicitly

File: components/articles/cards/ArticleCard.js [hunk 3]
```diff
@@ -42,45 +42,80 @@ const ArticleCard = props => {
...
-            <span className={linkCn({ dark })}>
-              <Text id="article.read-article" />
-            </span>
+        {!collection && (
+          <div className="article-card__description-container">
+            <div className="article-card__description">{subtitle}</div>
+            <div className={linkCn({ className: 'article-card__label-read', dark })}>
+              {type === 'text' && <Text id="article.read" />}
+              {type === 'audio' && <Text id="article.listen" />}
+              {type === 'video' && <Text id="article.watch" />}
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]:
```suggestion
	<Text id={LABEL_BY_TYPE[type]} />
```

File: components/articles/cards/ArticleCard.js [hunk 4]
```diff
@@ -42,45 +42,80 @@ const ArticleCard = props => {
...
+          </div>
+        )}
+        <div className="article-card__filler article-card__filler--top" />
+        <div className="article-card__title">
+          {type === 'video' && (
+            <Icon className="article-card__interactive-icon" pack="b" name="youtube" />
+          )}
+          {type === 'audio' && (
+            <Icon className="article-card__interactive-icon" pack="s" name="volume-up" />
+          )}
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]:
```js
const ICONS_BY_TYPE = {
	video: { pack: 'b', name: 'youtube' },
	audio: { pack: 's', name: 'volume-up' },
}

...

{ICONS_BY_TYPE[type] && <Icon className="article-card__interactive-icon" {...ICONS_BY_TYPE[type]}/>}
```

File: components/articles/cards/ArticleCard.js [hunk 5]
```diff
@@ -42,45 +42,80 @@ const ArticleCard = props => {
...
+              {type === 'text' && <Text id="article.read" />}
+              {type === 'audio' && <Text id="article.listen" />}
+              {type === 'video' && <Text id="article.watch" />}
+            </div>
+          </div>
+        )}
+        <div className="article-card__filler article-card__filler--middle" />
+        <div className="article-card__author-brand">
+          {brand && (
+            <img
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: why `brand` isn't tag link?

File: components/articles/cards/ArticleCard.js [hunk 6]
```diff
@@ -42,45 +42,80 @@ const ArticleCard = props => {
...
+            />
+          )}
+          {renderNodeList(authors.map(renderTag))}
+        </div>
+        <div className="article-card__filler article-card__filler--bottom" />
+        <div className="article-card__label-read-full">
+          <span className={linkCn({ dark })}>
+            {type === 'text' && <Text id="article.read-article" />}
+            {type === 'audio' && <Text id="article.listen-podcast" />}
+            {type === 'video' && <Text id="article.watch-video" />}
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: same as above

File: components/articles/cards/ArticleCard.js (hunk 3 above)
Comment (Drapegnik) [reply components/articles/cards/ArticleCard.js]: or even 
```js
<Text id={`article.${ACTION_BY_TYPE[type]}`} />
```

File: components/articles/cards/TagCard.js [hunk 7]
```diff
@@ -11,7 +11,7 @@ import { TOPIC } from 'constants/misc';
 import CardWrapper, { SIZES } from './CardWrapper';
 
 const TagCard = ({ slug, topicSlug, content, theme, size }) => {
-  const dark = theme === 'dark';
+  const dark = theme === 'dark' || slug === 'bssr' || slug === 'dali';
```
Comment (Drapegnik) [inline components/articles/cards/TagCard.js]: temp?

File: pages/admin/preview.js [hunk 8]
```diff
@@ -1,19 +1,25 @@
...
 
+import useBoolean from 'hooks/useBoolean';
+
 const mapStateToProps = (state, { lang }) => ({
   article: adminArticlesSelectors.getPreview(state, lang),
   error: adminArticlesSelectors.getError(state),
 });
 
 const ArticlePreview = ({ article, error }) => {
+  const [cardsPreview, togglePreview, _] = useBoolean(true);
```
Comment (Drapegnik) [inline pages/admin/preview.js]:
```suggestion
  const [cardsPreview, togglePreview] = useBoolean(true);
```

File: pages/admin/preview.js [hunk 9]
```diff
@@ -22,7 +28,25 @@ const ArticlePreview = ({ article, error }) => {
...
+    <>
+      <div className="preview-page__switch">
+        <input type="checkbox" checked={cardsPreview} onChange={togglePreview} />
+        {'  '}
+        <span>Preview Cards instead of the Article</span>
+      </div>
+      {cardsPreview && (
+        <div className="preview-page__cards" style={{ margin: '50px' }}>
+          {['xxl', 'xl', 'l', 'm', 'square-m', 'square-s'].map(size => (
+            <div className="preview-page__card">
```
Comment (Drapegnik) [inline pages/admin/preview.js]:
```suggestion
            <div key={size} className="preview-page__card">
```

File: pages/admin/preview.js [hunk 10]
```diff
@@ -22,7 +28,25 @@ const ArticlePreview = ({ article, error }) => {
...
+  return (
+    <>
+      <div className="preview-page__switch">
+        <input type="checkbox" checked={cardsPreview} onChange={togglePreview} />
+        {'  '}
+        <span>Preview Cards instead of the Article</span>
+      </div>
+      {cardsPreview && (
+        <div className="preview-page__cards" style={{ margin: '50px' }}>
+          {['xxl', 'xl', 'l', 'm', 'square-m', 'square-s'].map(size => (
```
Comment (Drapegnik) [inline pages/admin/preview.js]: use constant

**Review by Drapegnik — CHANGES_REQUESTED** [review]

File: components/articles/Article.js (hunk 1 above)
Comment (UladBohdan) [reply components/articles/Article.js]: nice catch

File: components/articles/cards/ArticleCard.js (hunk 2 above)
Comment (UladBohdan) [reply components/articles/cards/ArticleCard.js]: what was the reason to have it on a first place?

File: components/articles/cards/TagCard.js (hunk 7 above)
Comment (UladBohdan) [reply components/articles/cards/TagCard.js]: yup

File: components/articles/cards/ArticleCard.js (hunk 5 above)
Comment (UladBohdan) [reply components/articles/cards/ArticleCard.js]: fixed that

File: components/articles/cards/ArticleCard.js (hunk 2 above)
Comment (Drapegnik) [reply components/articles/cards/ArticleCard.js]: 3 different `<CardWrapper` with some shared props

File: components/articles/cards/ArticleCard.js [hunk 11]
```diff
@@ -14,37 +19,73 @@ import { linkCn } from 'utils/ui';
...
+  xl: { type: 'vertical', width: 313 },
+  l: { type: 'vertical', width: 240 },
+  m: { type: 'vertical', width: 186 },
+  'square-m': { type: 'horizontal', width: 390 },
+  'square-s': { type: 'horizontal', width: 300 },
+};
+
+// Q: How do I ensure width is a number here?
+const getCoverLink = (images, cardSize) =>
+  `${images[COVER_BY_CARD_SIZE[cardSize].type]}?w=${COVER_BY_CARD_SIZE[cardSize].width * 2}`;
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]:
```js
const getCoverLink = (images, cardSize) => {
	const { type, width } = COVER_BY_CARD_SIZE[cardSize];
	return `${images[type]}?w=${+width * 2}`;
}
```

Comment (Drapegnik) [reply components/articles/cards/ArticleCard.js]: TODO: separate `1x` & `2x`

File: components/articles/cards/ArticleCard.js [hunk 12]
```diff
@@ -76,46 +121,49 @@ const ArticleCard = props => {
...
-              className="article-card__brand"
-              src={brand.content.image}
-              alt={brand.content.title}
-            />
+            <Link
+              key={brand.slug}
+              route={ROUTES_NAMES.tag}
+              params={{ topic: brand.topicSlug, tag: brand.slug }}
+            >
+              <img
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: `<Image />`

File: components/articles/cards/ArticleCard.js [hunk 13]
```diff
@@ -55,17 +96,21 @@ const ArticleCard = props => {
...
+              (acc, [screenName, cardSize]) => {
+                acc[screenName] = getCoverLink(images, cardSize);
+                return acc;
+              },
+              {}
+            )}
+            alt={title}
+          />
+        ) : (
+          <img className="article-card__cover" src={getCoverLink(images, size)} alt={title} />
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: `<Image/>` with 1/2x

File: components/articles/cards/ArticleCard.js [hunk 14]
```diff
@@ -14,37 +19,73 @@ import { linkCn } from 'utils/ui';
...
+    images,
     title,
     type,
     slug,
     tagsByTopic,
   } = props;
   const dark = theme === 'dark';
   const {
-    brands: [brand],
+    brands: [brand], // Q: what about multiple brands.
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: just skip for now (YAGNI)

File: components/articles/cards/ArticleCard.js [hunk 15]
```diff
@@ -125,6 +173,13 @@ const ArticleCard = props => {
 
 ArticleCard.propTypes = {
   size: PropTypes.oneOf(SIZES),
+  // Q: I an not sure if the 'context' works as required.
+  context: PropTypes.oneOfType([
+    // In case card size is set explicitly, no context needs to be provided.
+    PropTypes.oneOf(['no']),
+    // In case card size is 'auto', the context (e.g. the block and position) must be provided.
+    PropTypes.arrayOf(PropTypes.string),
+  ]).isRequired,
```
Comment (Drapegnik) [inline components/articles/cards/ArticleCard.js]: just:
```js
context: PropTypes.arrayOf(PropTypes.string)
```

File: pages/admin/preview.js [hunk 16]
```diff
@@ -37,9 +38,9 @@ const ArticlePreview = ({ article, error }) => {
       </div>
       {cardsPreview && (
         <div className="preview-page__cards" style={{ margin: '50px' }}>
-          {['xxl', 'xl', 'l', 'm', 'square-m', 'square-s'].map(size => (
-            <div className="preview-page__card">
-              <ArticleCard {...article} size={size} />
+          {DEFAULT_SIZES.concat(SQUARE_SIZES).map(size => (
+            <div key={size} className="preview-page__card">
+              <ArticleCard {...article} size={size} context="no" />
```
Comment (Drapegnik) [inline pages/admin/preview.js]: omit `context="no"`

**Review by Drapegnik — CHANGES_REQUESTED** [review]

File: components/articles/Article.js [hunk 17]
```diff
@@ -19,6 +19,8 @@ import { ROUTES_NAMES } from 'routes';
 
 const COVER_SIZES = [1200, 1000, 770, 640, 360];
 
+const AUTHOR_BRAND_THUMBNAIL_WIDTH = 100;
```
Comment (Drapegnik) [inline components/articles/Article.js]: but shown as `48px`

**Comment by UladBohdan** [comment]
[to implement correct interpretation of dark/light theme for cards](https://trello.com/c/rtRK8jJE/413-to-implement-correct-interpretation-of-dark-light-theme-for-cards)

File: components/articles/Article.js (hunk 17 above)
Comment (UladBohdan) [reply components/articles/Article.js]: 48px is for square images. For some wide logos (like Art Museum) width is actually more than 48px. The height is 48px though

**Review by Drapegnik — APPROVED** [review]

**Pull request merged**